arrebol/worker/auth/auth.go

```diff
@@ -0,0 +1,60 @@
+package auth
+
+import (
+	"encoding/json"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/service/errors"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/auth/token"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/auth/allowlist"
+	"github.com/ufcg-lsd/arrebol-pb/arrebol/worker/key"
+	"github.com/ufcg-lsd/arrebol-pb/crypto"
+)
+
+type Authenticator struct {
+	allowlist allowlist.AllowList
+}
+
+func NewAuth() *Authenticator {
+	allowlist := allowlist.NewAllowList()
+	var auth = Authenticator{
+		allowlist: allowlist,
+	}
+	return &auth
+}
+
+func (auth *Authenticator) Authenticate(signature []byte, worker *worker.Worker) (token.Token, error) {
+	data, err := json.Marshal(worker)
+	var token token.Token
+	if err != nil {
+		return token, err
+	}
+	publicKey, err := key.GetPublicKey(worker.ID)
+	if err != nil {
+		return token, err
+	}
+	err = crypto.Verify(publicKey, data, signature)
+	if err != nil {
+		return token, err
+	}
+	if contains := auth.allowlist.Contains(worker.ID); contains {
```

Comment:
i think we should not mix the concepts of authorization (backed by this white-list) and authentication

Reply:
Okay. I moved the allow list check to the authorization method